addons/minio/template/base/install.sh

```diff
@@ -247,7 +271,344 @@ function allow_pvc_resize() {
+function minio_migrate_fs_backend() {
+    local minio_replicas
+    minio_replicas=$(kubectl get --no-headers deploy minio -n "$MINIO_NAMESPACE" -o custom-columns=:.spec.replicas 2>/dev/null)
+    if [ -z "$minio_replicas" ] || [ "$minio_replicas" = "0" ] || ! minio_uses_fs_format ; then
+        return
+    fi
+
+    # we need to guarantee that the minio is working before starting to migrate. i have seen cases where due to the upgrade
+    # logic some pods take a while to come back up and, especially, receive an ip address. let's first make sure we can
+    # reach minio and then move forward. best case scenario we will move forward immediately, worst case scenario we fail
+    # because we can't reach minio (we would fail anyways).
+    local minio_service_ip
+    minio_service_ip=$(kubectl -n ${MINIO_NAMESPACE} get service minio | tail -n1 | awk '{ print $3}')
+    printf "Awaiting installed minio readiness\n"
+    if ! spinner_until 300 minio_ready "$minio_service_ip" ; then
+        bail "Minio API failed to report healthy"
+    fi
+
+
+    local minio_access_key
+    minio_access_key=$(kubernetes_secret_value "$MINIO_NAMESPACE" minio-credentials MINIO_ACCESS_KEY)
+    if [ -z "$minio_access_key" ]; then
+        bail "Failed to read minio access key"
+    fi
+
+    local minio_secret_key
+    minio_secret_key=$(kubernetes_secret_value "$MINIO_NAMESPACE" minio-credentials MINIO_SECRET_KEY)
+    if [ -z "$minio_secret_key" ]; then
+        bail "Failed to read minio access key"
+    fi
+
+    local minio_pod_ip
+    minio_pod_ip=$(kubectl get pods -n "$MINIO_NAMESPACE" -l app=minio --no-headers -o custom-columns=:.status.podIP | grep -v none | head -1)
+    if [ -z "$minio_pod_ip" ]; then
+        bail "Failed to determine minio pod ip address"
+    fi
+
+    minio_disable_minio_svc
+    if ! spinner_until 120 minio_svc_has_no_endpoints "$MINIO_NAMESPACE" "minio" ; then
+        minio_enable_minio_svc
+        bail "Timeout waiting for minio service to be decomissioned"
+    fi
+
+    if ! minio_create_fs_migration_deployment ; then
+        minio_enable_minio_svc
+        bail "Failed to start minio migration deployment"
+    fi
+
+    if ! migrate_object_store \
+        "$MINIO_NAMESPACE" \
+        "$minio_pod_ip:9000" \
+        "$minio_access_key" \
+        "$minio_secret_key" \
+        "minio-migrate-fs-backend" \
+        "$minio_access_key" \
+        "$minio_secret_key"
+    then
+        minio_enable_minio_svc
+        minio_destroy_fs_migration_deployment
+        bail "Failed to migrate data to minio migration deployment"
+    fi
+
+    # scale down minio and wait until it is out of service.
+    kubectl scale deployment -n "$MINIO_NAMESPACE" minio --replicas=0
+    if ! spinner_until 120 deployment_fully_updated minio "$MINIO_NAMESPACE"; then
+        minio_enable_minio_svc
+        minio_destroy_fs_migration_deployment
+        kubectl scale deployment -n "$MINIO_NAMESPACE" minio --replicas="$minio_replicas"
```

Comment:
i wonder if its worth making a separate function for this as you repeat it so many times. also would it make sense to append `|| true` to each line to do the best we can reverting?

Reply:
good idea. fixed.